pulsar-client-admin-shaded/pom.xml

```diff
@@ -119,7 +119,6 @@
                   <include>org.reactivestreams:reactive-streams</include>
                   <include>com.typesafe.netty:netty-reactive-streams</include>
                   <include>org.javassist:javassist</include>
-                  <include>com.google.protobuf:protobuf-java</include>
```

Comment:
It should be with optional scope, right?

Reply:
> It shouldn't be included in the pulsar-client-all shaded jar either.

@lhotari Thanks for point it out, applied same change for `pulsar-client-all` component.

Reply:
> It should be with optional scope, right?

@merlimat 
The reduced `pom.xml` generated by the `maven-shade-plugin` lists `protobuf-java` as a `compile` scope dependency.

After reviewing the entire codebase, I found that the only client-side feature relying on `protobuf-java` is the `protobuf schema`. Therefore, it seems reasonable to make this dependency optional if `maven-shade-plugin` supports that configuration.

_Search entire codebase to find classes depends `protobuf-java`:_
```shell
# grep -ar 'import com.google.protobuf' **/*.java | grep -vE '^pulsar-broker|pulsar-functions|^managed-ledger|^pulsar-io'
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchema.java:import com.google.protobuf.Descriptors;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchema.java:import com.google.protobuf.GeneratedMessageV3;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaUtils.java:import com.google.protobuf.Descriptors;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufSchema.java:import com.google.protobuf.Descriptors;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufSchema.java:import com.google.protobuf.GeneratedMessageV3;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericProtobufNativeReader.java:import com.google.protobuf.Descriptors;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericProtobufNativeReader.java:import com.google.protobuf.DynamicMessage;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericProtobufNativeReader.java:import com.google.protobuf.InvalidProtocolBufferException;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericProtobufNativeRecord.java:import com.google.protobuf.Descriptors;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericProtobufNativeRecord.java:import com.google.protobuf.DynamicMessage;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericProtobufNativeSchema.java:import com.google.protobuf.Descriptors;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/MultiVersionGenericProtobufNativeReader.java:import com.google.protobuf.Descriptors;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/ProtobufNativeRecordBuilderImpl.java:import com.google.protobuf.Descriptors;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/ProtobufNativeRecordBuilderImpl.java:import com.google.protobuf.DynamicMessage;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/reader/ProtobufReader.java:import com.google.protobuf.InvalidProtocolBufferException;
pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/reader/ProtobufReader.java:import com.google.protobuf.Parser;
pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaTest.java:import com.google.protobuf.Descriptors;
pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaUtilsTest.java:import com.google.protobuf.Descriptors;
pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/generic/GenericProtobufNativeReaderTest.java:import com.google.protobuf.Descriptors;
pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/generic/GenericProtobufNativeReaderTest.java:import com.google.protobuf.DynamicMessage;
pulsar-metadata/src/main/java/org/apache/pulsar/metadata/bookkeeper/PulsarLedgerUnderreplicationManager.java:import com.google.protobuf.InvalidProtocolBufferException;
pulsar-metadata/src/main/java/org/apache/pulsar/metadata/bookkeeper/PulsarLedgerUnderreplicationManager.java:import com.google.protobuf.TextFormat;
pulsar-metadata/src/test/java/org/apache/pulsar/metadata/bookkeeper/LedgerUnderreplicationManagerTest.java:import com.google.protobuf.TextFormat;
tests/docker-images/java-test-functions/src/main/java/org/apache/pulsar/tests/integration/io/TestGenericObjectSink.java:import com.google.protobuf.DynamicMessage;
```

Reply:
> After reviewing the entire codebase, I found that the only client-side feature relying on `protobuf-java` is the `protobuf schema`. Therefore, it seems reasonable to make this dependency optional if `maven-shade-plugin` supports that configuration.

I doubt that it could be optional for pulsar-client-all or pulsar-client shaded clients since there's a hard reference to Protobuf classes in `org.apache.pulsar.client.api.Schema` interface: https://github.com/apache/pulsar/blob/5d6a88efa78969073c9dad015b4727b59e8a76d8/pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Schema.java#L280-L289 

For pulsar-admin-client shaded, it's ok to exclude it since there aren't classes referencing Protobuf generated classes in pulsar-admin-client.

Reply:
I guess removing `extends com.google.protobuf.GeneratedMessageV3` could be a way to handle it, but that would be a breaking change in the API.

Reply:
If user not use `protobuf schema`, it's ok to do that. According to [Java Language Specification][jls] JVM only initialize classes that actually going to use.

[jls]: https://docs.oracle.com/javase/specs/jls/se17/html/jls-12.html#jls-12.2:~:text=12.4.1.-,When%20Initialization%20Occurs,-A%20class%20or

Reply:
Making the schema dependency `optional` helps reduce unnecessary transitive dependencies (e.g., protobuf, avro). If users need to use protobuf or avro schemas, they can still manually add the dependency later. However, this could introduce a breaking change for users who are already relying on those dependencies being present.

IMO, making `protobuf-java` optional is beyond the scope of this PR. It’s something worth discussing separately at a later stage.

Reply:
> If user not use `protobuf schema`, it's ok to do that. According to [Java Language Specification](https://docs.oracle.com/javase/specs/jls/se17/html/jls-12.html#jls-12.2:~:text=12.4.1.-,When%20Initialization%20Occurs,-A%20class%20or) JVM only initialize classes that actually going to use.

Yes, that could apply in this case. The fact that the Protobuf type is referenced in a type parameter helps a lot. However, that could cause class loading issues in reflection based libraries that run introspection to find out type parameters, but that's a very rare case.

Reply:
> IMO, making `protobuf-java` optional is beyond the scope of this PR. It’s something worth discussing separately at a later stage.

makes sense, let's handle that separately.

Reply:
I'll mark this conversation resolved since the PR LGTM for it to be merged.

Reply:
> IMO, making protobuf-java optional is beyond the scope of this PR. It’s something worth discussing separately at a later stage.

It was always meant to be optional. It's already optional in the unshaded module. It was included as required dependency by mistake and this is why it's getting included in the shaded jar. 

It don't think it's beyond the scope to fix the root cause of the issue.

Reply:
@merlimat I created #23469 to track the issue separately so that we can target that with another PR. /cc @Shawyeok